File: _algorithms_challenges/projecteuler/ProjectEuler-master(2)/ProjectEuler-master/3/3.pl
use strict;
use warnings;

sub sieve {
    my $bound = shift;
    my @primes = ();
    my @visited = ();
    for (my $i = 2; $i <= $bound; $i++) {
        next if $visited[$i];
        push @primes, $i;
        for (my $j = $i + $i; $j <= $bound; $j += $i) {
            $visited[$j] = 1;
        }
    }
    return @primes;
}

sub get_largest_prime_factor {
    my $number = shift;
    my $bound = int(sqrt $number) + 1;
    my @primes = reverse sieve($bound);
    foreach my $factor (@primes) {
        if ($number % $factor == 0) {
            return $factor;
        }
    }
    return $number;
}

sub main {
    print get_largest_prime_factor(13195) . "\n";
    print get_largest_prime_factor(600851475143) . "\n";
}

main();

File: _algorithms_challenges/projecteuler/ProjectEuler-master(2)/ProjectEuler-master/3/3.py
import sys

def main():
    primes = set(range(2, 780000))
    for i in range(2, 780000): 
        primes -= set(range(2*i, 780000, i))
    
    d = 600851475143
    factor = []
    for p in primes:
        while (d % p == 0):
            d /= p
            factor += [p]
    if d != 1:
        factor += [int(d)]
        
    print(factor[-1])
    
if __name__ == '__main__':
    sys.exit(main())
